src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2047,7 +2047,376 @@ cdef class Polynomial(CommutativePolynomial):
+            if not hd.is_zero() and hd != self.degree():
+                if 2*hd <= self.degree():
+                    return h._cantor_zassenhaus_split_to_irreducible(degree)
+                else:
+                    return (self//h)._cantor_zassenhaus_split_to_irreducible(degree)
+
+        # If you are reaching this error, chances are there's a bug in the code.
+        raise AssertionError(f"no splitting of degree {degree} found for {self}")
+
+    def _any_irreducible_factor_squarefree(self, degree=None):
+        """
+        Helper function for any_irreducible_factor which computes
+        an irreducible factor from self, assuming the input is
+        squarefree.
+
+        Does this by first computing the distinct degree factorisations
+        of self and thenfinds a factor with Cantor-Zassenhaus
+        splitting.
+
+        If degree is not None, then only irreducible factors of degree
+        `degree` are searched for, otherwise the smallest degree factor
+        is found.
+
+        EXAMPLES::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F = GF(163)
+            sage: R.<x> = F[]
+            sage: f = (x^2 + 36*x + 34) * (x^2 + 36*x + 132) * (x^2 + 112*x + 48)
+            sage: f = (x^2 + 50*x + 28) * (x^5 + 75*x^4 + 87*x^3 + 96*x^2 + 98*x + 40) * (x^7 + 20*x^6 + 16*x^5 + 36*x^4 + 59*x^3 + 3*x^2 + 46*x + 84)
+            sage: f._any_irreducible_factor_squarefree()
+            x^2 + 50*x + 28
+            sage: f._any_irreducible_factor_squarefree(degree=2)
+            x^2 + 50*x + 28
+            sage: f._any_irreducible_factor_squarefree(degree=5)
+            x^5 + 75*x^4 + 87*x^3 + 96*x^2 + 98*x + 40
+            sage: f._any_irreducible_factor_squarefree(degree=7)
+            x^7 + 20*x^6 + 16*x^5 + 36*x^4 + 59*x^3 + 3*x^2 + 46*x + 84
+            sage: g = (x^2 + 50*x + 28)
+            sage: g._any_irreducible_factor_squarefree(degree=1)
+            Traceback (most recent call last):
+            ...
+            ValueError: no irreducible factor of degree 1 could be computed from x^2 + 50*x + 28
+
+        ::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F.<z4> = GF(2^4)
+            sage: R.<x> = F[]
+            sage: f = (x + z4^3 + z4^2 + z4) * (x^2 + x + z4^3 + 1) * (x^3 + (z4^3 + z4)*x^2 + z4^2 + 1)
+            sage: f._any_irreducible_factor_squarefree()
+            x + z4^3 + z4^2 + z4
+            sage: f._any_irreducible_factor_squarefree(degree=1)
+            x + z4^3 + z4^2 + z4
+            sage: f._any_irreducible_factor_squarefree(degree=2)
+            x^2 + x + z4^3 + 1
+            sage: f._any_irreducible_factor_squarefree(degree=3)
+            x^3 + (z4^3 + z4)*x^2 + z4^2 + 1
+        """
+        # If the degree is not None we only want to check a single polynomial
+        if degree is not None:
+            for (poly, d) in self._distinct_degree_factorisation_squarefree():
+                if d == degree and not poly.is_one():
```

Comment:
I suppose if `_distinct_degree_factorisation_squarefree()` is adjusted to never return a tuple with `poly == 1` (see above) then this check can be simplified.